### PR TITLE
[WIP] Add html_repr's

### PR DESCRIPTION
**Pull request by Tillsten**
Voices from the dead again :-)

unbenannt

I try to make the output of my software a bit little nicer to look at. One thing it is missing is a nice html representation of the fit results. I think this may fit directly into lmfit. There are different approaches to choose from:

* This PR simply adds a html_repr method to parameters, which will be used automatically in a notebook environment. While I prefer this approach, it may change the output of existing notebooks. 

* Add an extra `print_html` function. This would require the end-user to use the ipython `display` function manually, which I think is not very user-friendly. 

* Combine the first approach with some method of configuration, either via a lmfit setting/function or some environments. This would introduce state to lmfit, but it may be worth it. 

So is there any interest at all? If so, which approach to we want to choose? Of course the `Result`-class and the CI classes can get similar methods, making the use of the report functions in the notebook unnecessary for standard usage.

### CODE REVIEW

**Comment by newville** [comment]
@Tillsten Great to hear from you!  I think this is a fine idea, though I'm not sure i really know all the subtleties or options:  I'm remarkably ignorant of all things Jupyter.  Is it enough to just have a `_repr_html_()` method that is used by Jupyter to overwrite `_repr_()`?    Or, is there a way to identify that one is running Jupyter and then dynamically alter (aka monkey-patch) `_repr_` to be `_repr_html_`?

**Comment by tritemio** [comment]
@newville, jupyter notebook uses `_repr_html_()` if available, otherwise falls back to `__repr__()`. No need to overwrite anything. Define `_repr_html_()` and it will just work.

**Comment by newville** [comment]
@tritemio @Tillsten  Ah, that's even better and makes perfect sense.   And then, Yes to adding `_repr_html_()` methods wherever it is appropriate and useful.  An HTML table of Parameters is definitely one such place.

**Comment by reneeotten** [comment]
would it be possible to put the HTML-generating code into ```printfuncs.py``` and point the ```_repr_html_()``` methods to it? That would, in my opinion, avoid cluttering these HTML generating functions throughout the lmfit codebase.

**Comment by newville** [comment]
@reneeotten @Tillsten  should we merge this and then consider modifying as Renee suggests?   It seems like a useful feature.

It looks to me like the test failure is actually not from the code added here but from other problems with assumptions about the testing environment for "Py27 + version=minimum".

**Comment by Tillsten** [comment]
I added also a repr for the MinimizerResult but it is not yet tested if someone has time. I am ambivalent about where the code should be, there are arguments for both: in the class, since the code is useless without it, and in printfucns.py, since there is already some printing code.

**Comment by newville** [comment]
@Tillsten  Thanks!  I think the `MinimizerResult` html repr will be very nice.
I do share your ambivalence about where the bulk of this code should go, but I'm also willing to defer to @reneeotten on this.   Moving the code might allow for some generalization of "html reprs" so that (for example) the Confidence Interval report could also have an HTML/Jupyter version even if that is not exactly a "repr".   Anyway, these might all be helpful for the Jupyter widgets that @wholden has been working on (https://github.com/wholden/jupyter-lmfit).

Is it OK with you and @reneeotten if we merge this and then decide if code should be moved?

File: lmfit/parameter.py [hunk 1]
```diff
@@ -284,6 +284,53 @@ def pretty_print(self, oneline=False, colwidth=8, precision=4, fmt='g',
...
+        # Helper functions for shorter code
+        html = []
+        add = html.append
+        cell = lambda x: add('<td>%s</td>' % x)
+        head = lambda x: add('<th>%s</th>' % x)
+
+        add('<table>')
+        # Header ----
+        add('<tr>')
+        headers = ['name', 'value', '(init)', 'min', 'max', 'vary']
```
Comment (reneeotten) [inline lmfit/parameter.py]: please add ```brute_step``` as well, as in ```pretty_print()```

File: lmfit/parameter.py [hunk 2]
```diff
@@ -284,6 +284,53 @@ def pretty_print(self, oneline=False, colwidth=8, precision=4, fmt='g',
...
+            cell('%f' % p.value)
+            if p.stderr is not None:
+                cell('+/-')
+                cell('%f' % p.stderr)
+                cell('%.2f%%' % (100 * p.stderr / p.value))
+            elif any_err:
+                cell('')
+                cell('')
+                cell('')
+            cell(p.user_value)
```
Comment (reneeotten) [inline lmfit/parameter.py]: this causes problems, in the example below:
```
import lmfit
p = lmfit.Parameters()
p.add_many(('a1', 4), ('a2', 4), ('t1', 3), ('t2', 3., True))

p
```

with the following traceback

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    343             method = get_real_method(obj, self.print_method)
    344             if method is not None:
--> 345                 return method()
    346             return None
    347         else:

~/Library/Python/3.7/lib/python/site-packages/lmfit-0.9.12+82.gd464830-py3.7.egg/lmfit/parameter.py in _repr_html_(self)
    320                 cell('')
    321                 cell('')
--> 322             cell(p.user_value)
    323             cell('%f' % p.min)
    324             cell('%f' % p.max)

AttributeError: 'Parameter' object has no attribute 'user_value'
````

**Comment by Tillsten** [comment]
Sorry I have too much on the table to tackle the last changes, thanks for taking over Matt!

Closed for #548.

**Comment by newville** [comment]
@Tillsten No problem! Thanks for the suggestion and start of the code to add the html reprs.

**Pull request closed without merging**